arbitrator/jit/src/gostack.rs

```diff
@@ -2,35 +2,74 @@
 // For license information, see https://github.com/nitro/blob/master/LICENSE
 
 use crate::{
-    machine::{WasmEnv, WasmEnvArc},
+    machine::{WasmEnv, WasmEnvMut},
     syscall::JsValue,
 };
 
-use parking_lot::MutexGuard;
+use ouroboros::self_referencing;
 use rand_pcg::Pcg32;
-use wasmer::{Memory, MemoryView, WasmPtr};
+use wasmer::{AsStoreRef, Memory, MemoryView, StoreRef, WasmPtr};
 
 use std::collections::{BTreeSet, BinaryHeap};
 
-#[derive(Clone)]
+#[self_referencing]
+struct MemoryViewContainer {
+    memory: Memory,
+    #[borrows(memory)]
+    #[covariant]
+    view: MemoryView<'this>,
+}
+
+impl MemoryViewContainer {
+    fn create(env: &WasmEnvMut<'_>) -> Self {
+        // this func exists to properly constrain the closure's type
+        fn closure<'a>(
+            store: &'a StoreRef,
+        ) -> impl (for<'b> FnOnce(&'b Memory) -> MemoryView<'b>) + 'a {
+            move |memory: &Memory| memory.view(&store)
+        }
+
+        let store = env.as_store_ref();
+        let memory = env.data().memory.clone().unwrap();
+        let view_builder = closure(&store);
+        MemoryViewContainerBuilder {
+            memory,
+            view_builder,
+        }
+        .build()
+    }
+
+    fn view(&self) -> &MemoryView {
+        self.borrow_view()
```

Comment:
What if we just impl'd Deref for MemoryView so you could use it as a view?

Reply:
That sounds awesome but unfortunately the impl isn't possible
```rs
impl Deref for MemoryViewContainer {
    type Target = MemoryView<'a>;

    fn deref(&'a self) -> &MemoryView<'a> {
        self.borrow_view()
    }
}
```
for an `'a` we can't express